File: MDP/MDPNoAdversary.cpp
/* Marmote and MarmoteMDP are free softwares: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

Marmote is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with Marmote. If not, see <http://www.gnu.org/licenses/>.

Copyright 2022 Emmanuel Hyon, Alain Jean-Marie*/


#include "marmoteCore/marmoteInterval.h"
#include "marmoteCore/marmoteSparseMatrix.h"
#include "marmoteCore/marmoteBox.h"
#include "marmoteMDP/marmoteAverageMDP.h"
#include "marmoteMDP/marmoteFeedbackSolutionMDP.h"
#include "marmoteMDP/marmoteSolutionMDP.h"

#include <list>
#include <vector>
#include <string>
#include <fstream>
#include <iostream>

using namespace std;
/**
 * Class to impliment MDP for autoscaling decision without adversary present based on Tournaire model
 * @author Chamberlain, Jonathan
 * @version 4
 * @date January 2024
 * 
 */
int main( int argc, char** argv )
{
    
    string critere("min");
        
    //create and initialize epsilon, delta.
    double epsilon = 0.000001;
    double delta = 0.000001;

    //create and initialize the maximum number of iterations allowed.
    int maxIter = 1000;

    // define costs
    double CA = 2; // activation cost
    double CD = 2; // deactiavtion cost
    double CS = 1; // cost of operating each node
    double CH = 1; // cost of holding each request
    double CR = 5; // cost of dropped requests
    double INF = 100000000; // arbitrary cost for invalid action

    // Create MDP object
    
    // define parameters for states to iterate on with MarmoteBox
    int n, r; // indices indicating number of nodes, number of requests
    int B = 100; // maximum buffer for number of requests
    int N = 16; // maximum number of nodes
    stateType indexO; // index of current/origin state 
    stateType indexD; // index of destination state 
    stateType k; // ordinal indicies of states
    // create box such that indcies k map to states of type (n,r)
    stateType dims[2]={N,B+1}; // c++ is zero indexed.
    MarmoteBox *stateSpace = new MarmoteBox(2,dims);

    // define transition rates
    double lambda = 50;
    double mu = 100;
    double LAM = lambda + N*mu; // maximum transition rate, used for normalization
    double p,q; // probabilities of arrivals, departures

    // three actions, subtract node (0), maintain nodes (1), add node (2)
    MarmoteSet *actionSpace = new MarmoteInterval(0,2);    
    
   
     // compute size of cardinals for state, action spaces
    stateType dim_SS = stateSpace->Cardinal();
    stateType dim_AS = actionSpace->Cardinal();
    // Allocate buffers for state spaces
    // manage origin/pre-transition state
    MarmoteState statebufferO = stateSpace->StateBuffer();
    // manage destination/post-transition state
    MarmoteState statebufferD = stateSpace->StateBuffer();

    // use an iterator on the action space
    MarmoteState actionbuffer = actionSpace->StateBuffer();
    
    vector<TransitionStructure*> trans(dim_AS);
    
    // Fill in transtion probailities for each action

    // Action 0 - subtract node; invalid action if already at minimum number of nodes
    actionSpace->FirstState(actionbuffer);
    SparseMatrix *P0 = new SparseMatrix(dim_SS); 
    // skip index n=0 as cannot subtract nodes when at minimum
    for(n = 1; n < N; n++) {
        statebufferO[0]=n; 
        statebufferD[0] = n-1;
        for(r = 0; r <= B ; r++) {
            // skip index n=0 as cannot subtract nodes when at minimum
            statebufferO[1]=r;
            indexO=stateSpace->Index(statebufferO);
            if (r < B) {
                // arrival case
                p = lambda/LAM;
                statebufferD[1] = r+1;
                indexD = stateSpace->Index(statebufferD);
                P0->addToEntry(indexO,indexD,p);
            } else {
                p = 0; // no arrival possible, handle self transition appropriately
            }
            if (r > 0) {
                // depature case
                // service rate depends on requests, number of nodes after subtracting; number of nodes off by 1 due to indexing
                q = mu*std::min(n,r)/LAM;  
                statebufferD[1] = r-1;
                indexD = stateSpace->Index(statebufferD);
                P0->addToEntry(indexO,indexD,q);
            } else {
                q = 0; // no departure possible, handle self transition appropriately
            }
            // psuedo-event self transition
            P0->addToEntry(indexO,indexO,1-p-q);
        }
    }
    trans.at(actionSpace->Index(actionbuffer)) = P0;

    // Action 1 - maintain nodes; always valid, creates B-D chain at each level
    actionSpace->NextState(actionbuffer);
    SparseMatrix *P1 = new SparseMatrix(dim_SS);
    for(n = 0; n < N; n++) {
        statebufferO[0] = n; 
        statebufferD[0] = n;
        for(r = 0; r <= B ; r++) {
            statebufferO[1]=r;
            indexO=stateSpace->Index(statebufferO);
            if (r < B) {
                // arrival case
                p = lambda/LAM;
                statebufferD[1] = r+1;
                indexD = stateSpace->Index(statebufferD);
                P1->addToEntry(indexO,indexD,p);
            } else {
                p = 0; // no arrival possible, handle self transition appropriately
            }
            if (r > 0) {
                // depature case
                // service rate depends on requests, number of nodes; number of nodes off by 1 due to indexing
                q = mu*std::min(n+1,r)/LAM; 
                statebufferD[1] = r-1;
                indexD = stateSpace->Index(statebufferD);
                P1->addToEntry(indexO,indexD,q);
            } else {
                q = 0; // no departure possible, handle self transition appropriately
            }
            // psuedo-event self transition
            P1->addToEntry(indexO,indexO,1-p-q);
        }
    }  
    trans.at(actionSpace->Index(actionbuffer)) = P1;
    
    // Action 2 - add node; invalid if already at max nodes
    actionSpace->NextState(actionbuffer);
    SparseMatrix *P2 = new SparseMatrix(dim_SS);
    // skip index n=N-1 as cannot add nodes when at maximum 
    for (n = 0; n < N-1; n++) {
        statebufferO[0] = n;
        statebufferD[0] = n+1;
        for (r = 0; r <= B ; r++) {
            statebufferO[1]=r;
            indexO=stateSpace->Index(statebufferO);
            if (r < B) {
                // arrival case
                p = lambda/LAM;
                statebufferD[1] = r+1;
                indexD = stateSpace->Index(statebufferD);
                P2->addToEntry(indexO,indexD,p);
            } else {
                p = 0; // no arrival possible, handle self transition appropriately
            }
            if (r > 0) {
                // depature case
                // service rate depends on requests, number of nodes after adding; number of nodes off by 1 due to indexing
                q = mu*std::min(r,n+2)/LAM; 
                statebufferD[1] = r-1;
                indexD = stateSpace->Index(statebufferD);
                P2->addToEntry(indexO,indexD,q);
            } else {
                q = 0; // no departure possible, handle self transition appropriately
            }
            // psuedo-event self transition
            P2->addToEntry(indexO,indexO,1-p-q);
        }
    }
    trans.at(actionSpace->Index(actionbuffer)) = P2;
    
    // define rewards; done in terms of running costs, activation costs, deactivation costs, holding costs
    // arbitrary large cost defined for invalid actions

    // Reward = CA*(activation == true)+CD*(deactivation==true)+CS*(numactiveNodes)+CH*(numreqs)
    SparseMatrix *Reward  = new SparseMatrix(dim_SS,dim_AS);
    /*
    (Normalized) Operating cost is equal to
    (CA*(activation == true) + CD*(deactivation == true)) +
    (CR*(r == B)) * lambda/LAM +
    (CS*N(n,a) + CH*r)

    where N(n,a) is the number of machines after taking action a when starting with nodes n
    */
   
    for (n = 0; n < N; n++) {
        statebufferO[0] = n;
        // handle general case where buffer not yet full
        for (r = 0; r < B; r++) {
            statebufferO[1] = r;    
            indexO = stateSpace->Index(statebufferO);
            // maintain number of nodes; always valid; add 1 to number of nodes to account for 0 indexing
            Reward->addToEntry(indexO,1,(n+1)*CS+r*CH);
            if (n > 0) {
                // more than 1 node, can subtract
                Reward->addToEntry(indexO,0,CD+n*CS+r*CH);
            } else {
                // subtraction invalid
                Reward->addToEntry(indexO,0,INF);
            }
            if (n < N-1) {
                // less than maximum number of nodes, can add
                Reward->addToEntry(indexO,2,CA+(n+2)*CS+r*CH);
            } else {
                // addition invalid
                Reward->addToEntry(indexO,2,INF);
            }
        }
        // special case where buffer full, must also account for dropped requests
        statebufferO[1] = B;
        indexO = stateSpace->Index(statebufferO);
        // maintain number of nodes; always valid; add 1 to number of nodes to account for 0 indexing
        Reward->addToEntry(indexO,1,(n+1)*CS+r*CH+(lambda/LAM)*CR);
        if (n > 0) {
            // more than 1 node, can subtract
            Reward->addToEntry(indexO,0,CD+n*CS+r*CH+(lambda/LAM)*CR);
        } else {
            // subtraction invalid
            Reward->addToEntry(indexO,0,INF);
        }
        if (n < N-1) {
            // less than maximum number of nodes, can add
            Reward->addToEntry(indexO,2,CA+(n+2)*CS+r*CH+(lambda/LAM)*CR);
        } else {
            // addition invalid
            Reward->addToEntry(indexO,2,INF);
        }
    }

    fstream file;
    file.open("mdpsolution.txt", ios::out);
    cout.rdbuf(file.rdbuf());

    cout << "Size :\t" << stateSpace->Cardinal() << std::endl;
    cout << "Begining building of MDP" << std::endl;
    AverageMDP *mdp = new AverageMDP(critere, stateSpace, actionSpace, trans, Reward);
    cout << "End of building MDP" << std::endl;

    cout << "Writing MDP" << endl;
    mdp->WriteMDP();
    
    cout << endl << "Writing solution modified policy iteration" << endl ;
    //call the function to solve the MDP.
    SolutionMDP *optimum = mdp->PolicyIterationModified(epsilon, maxIter,delta,maxIter);
    optimum->WriteSolutionByDim(1,stateSpace);
        
    cout <<endl << endl << "Checking solutions" << endl ;
    double *sol = mdp->PolicyCost(optimum,epsilon, maxIter);
    for(int i=0;i<stateSpace->Cardinal();i++){
           cout <<"i= " << i << "  sol= " << sol[i] << endl;   
    }
    
    cout<< endl << endl << "********************************" << endl ;
        
    // cleanup
    file.close();    
    mdp->ClearRew();
    delete mdp;
    delete optimum;
    delete P0;
    delete P1;
    delete P2;
    delete[] sol;
    delete[] actionbuffer;

    return 0;
}

File: MDP/MDPNoAdversary.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

#Marmote and MarmoteMDP and pyMarmoteMDP are free softwares: you can redistribute it and/or modify
#it under the terms of the GNU General Public License as published by
#the Free Software Foundation, either version 3 of the License, or
#(at your option) any later version.

#Marmote is distributed in the hope that it will be useful,
#but WITHOUT ANY WARRANTY; without even the implied warranty of
#MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
#GNU General Public License for more details.

#You should have received a copy of the GNU General Public License
#along with MarmoteMDP. If not, see <http://www.gnu.org/licenses/>.

#Copyright 2019 Emmanuel Hyon, Alain Jean-Marie

"""
 @brief Class to implement MDP for Kubernetes autoscaling decision 
 @author jdchambo
 @date June 2024
 @version 0.1

 This MDP Considers a secnario where a microservice running on Kubernetes is subject to autoscaling based on the amount of traffic present

 This model has three actions:
  -1, remove a service unit
  0, keep number of service units the same
  1, add a service unit

 The number of states is variable and controlled by two variables:
    N, the limit on the number of Service Units
    B, a buffer for the maximum number of requests
"""

# import the library
import marmote.core as mc
import marmote.mdp as mmdp
import numpy as np

# We want to minimize the costs at each stage
critere = "min"
# here is the discount factor
#beta=0.95
#here are the parameters for the modified iteration
epsilon = 0.0001
# delta = 0.0001
maxIter = 1000

# Define costs (in cents) 
Ca =  2 # Activation Cost
Cd =  2 # Deactivation Cost
Cs = 5 # Service Unit operation Cost
Ch =  5 # Request Holding cost 
Cr =  10 # Cost of Dropped Request

# creating the state space
N = 16 # limit on number of Service Units
B = 100 # request Buffer
boxdims = np.array([N, B+1]) # dimensions of the Marmote Box defining the state space
stateSpace = mc.MarmoteBox(boxdims) # creates state space [0, ..., N-1] x [0, B] due to python indexing
dimSS = stateSpace.Cardinal()

# creating the action space as interval of the possible actions labeled 0, 1, 2
actionSpace = mc.MarmoteInterval(0,2)
dimAS = actionSpace.Cardinal()

def NumSUs(su,act):
    # determine the number of Service Units resulting from action
    new = min(max(1,su+act),N)
    return new

# transition rates
lam = 0.356 # arrival rate of requests, in requests per second; corresponds to ~ interarrival time of 2.81s
mu = 0.0003 # departure rate of requests, in requests per second; corresponds to mean serice time of ~ 0.95h or 3420s
LAM = lam + N*mu # maximum transition rate, used for normalization

# define Cost Matrix
CostMat = mc.FullMatrix(dimSS,dimAS)
etat = np.array([0,0]) # get initial state space
for k in range(dimSS):
    # compute state index
    indexO = stateSpace.Index(etat)
    n = etat[0]+1 # number of nodes; add 1 to offset 0 index
    r = etat[1] # requests
    # define the cost for each action
    if (r == B):
        # account for full buffer
        CostMat.setEntry(indexO,0,(Cd+NumSUs(n,-1)*Cs+r*Ch+lam*Cr)/LAM)
        CostMat.setEntry(indexO,1,(NumSUs(n,0)*Cs+r*Ch+lam*Cr)/LAM)        
        CostMat.setEntry(indexO,2,(Ca+NumSUs(n,1)*Cs+r*Ch+lam*Cr)/LAM)
    else:
        CostMat.setEntry(indexO,0,(Cd+NumSUs(n,-1)*Cs+r*Ch)/LAM)
        CostMat.setEntry(indexO,1,(Cs*NumSUs(n,0)+r*Ch)/LAM)
        CostMat.setEntry(indexO,2,(Ca+NumSUs(n,1)*Cs+r*Ch)/LAM)
    stateSpace.NextState(etat)

#print("********************************")
#print("Cost Matrix")
#print(CostMat)


# Compute transition value for each state.

#Create matrix corresponding to each action 
P0 = mc.SparseMatrix(dimSS) # action -1
P1 = mc.SparseMatrix(dimSS) # action 0
P2 = mc.SparseMatrix(dimSS) # action 1
trans = [P0,P1,P2]
etat = np.array([0,0]) # get initial state space
sortie = np.array([0,0]) # array to represent the end state following transition, intialize to dummy state
for k in range(dimSS):
    # compute state index
    indexO = stateSpace.Index(etat)
    n = etat[0] + 1 # number of nodes; add 1 to offset 0 index
    r = etat[1] # requests
    for a in range(3):
        act = a-1
        # arrivals
        if r < B:
            p = lam/LAM
            sortie[0] = NumSUs(n,act) - 1 #index offset by 1
            sortie[1] = r + 1
            indexD = stateSpace.Index(sortie)
            trans[a].setEntry(indexO,indexD,p)
        else:
            # special case full buffer, no arrivals
            p = 0
        # departures
        if r > 0:
            q = mu*min(r,NumSUs(n,act))/LAM
            sortie[0] = NumSUs(n,act) - 1 # index offset by 1
            sortie[1] = r - 1
            indexD = stateSpace.Index(sortie)
            trans[a].setEntry(indexO,indexD,q)
        else:
            # special case empty buffer, no depatures
            q = 0
        # self transition psuedo event
        trans[a].setEntry(indexO,indexO,1-p-q)
    stateSpace.NextState(etat)


#print("********************************")
#print("Probability Matrix Action = -1")
#print(P0)
#print("Probability Matrix Action = 0")
#print(P1)
#print("Probability Matrix Action = 1")
#print(P2)
#print("********************************") 
print("Building MDP")
mdp = mmdp.AverageMDP(critere, stateSpace, actionSpace, trans, CostMat)

print("Solving using Value Iteration")
optimum = mdp.ValueIteration(epsilon,maxIter)
print("********************************")
print("Printing Solution")
line = optimum.SolutionByDim(1,stateSpace)
print(line)

#print("Solving using modified Policy Iteration")
#call the function to solve the MDP.
#optimum2 = mdp.PolicyIterationModified(epsilon, maxIter, delta, maxIter)

#print("********************************")
#print("Printing Solution")
#line2 = optimum2.SolutionByDim(1,stateSpace)
#print(line2)
